lib/sync/service.go

```diff
@@ -0,0 +1,258 @@
+	ShowMetrics()
+}
+
+type BlockOrigin byte
+
+const (
+	networkInitialSync BlockOrigin = iota
+	networkBroadcast
+)
+
+type SyncService struct {
+	mu         sync.Mutex
+	wg         sync.WaitGroup
+	network    Network
+	blockState BlockState
+
+	currentStrategy Strategy
+	defaultStrategy Strategy
+
+	workerPool        *syncWorkerPool
+	waitPeersDuration time.Duration
+	minPeers          int
+	slotDuration      time.Duration
+
+	stopCh chan struct{}
+}
+
+func NewSyncService(network Network,
+	blockState BlockState,
+	currentStrategy, defaultStrategy Strategy) *SyncService {
+	return &SyncService{
+		network:           network,
+		blockState:        blockState,
+		currentStrategy:   currentStrategy,
+		defaultStrategy:   defaultStrategy,
+		workerPool:        newSyncWorkerPool(network),
+		waitPeersDuration: 2 * time.Second,
+		minPeers:          1,
+		stopCh:            make(chan struct{}),
+	}
+}
+
+func (s *SyncService) waitWorkers() {
+	waitPeersTimer := time.NewTimer(s.waitPeersDuration)
+	bestBlockHeader, err := s.blockState.BestBlockHeader()
+	if err != nil {
+		panic(fmt.Sprintf("failed to get highest finalised header: %v", err))
+	}
+
+	for {
+		total := s.workerPool.totalWorkers()
+		logger.Debugf("waiting peers...")
+		logger.Debugf("total workers: %d, min peers: %d", total, s.minPeers)
+		if total >= s.minPeers {
+			return
+		}
+
+		err := s.network.BlockAnnounceHandshake(bestBlockHeader)
+		if err != nil && !errors.Is(err, network.ErrNoPeersConnected) {
+			logger.Errorf("retrieving target info from peers: %v", err)
+		}
+
+		select {
+		case <-waitPeersTimer.C:
+			waitPeersTimer.Reset(s.waitPeersDuration)
+
+		case <-s.stopCh:
+			return
+		}
+	}
+}
+
+func (s *SyncService) Start() error {
+	s.waitWorkers()
+
+	s.wg.Add(1)
+	go s.runSyncEngine()
+	return nil
+}
+
+func (s *SyncService) Stop() error {
+	// TODO: implement stop mechanism
+	close(s.stopCh)
+	s.wg.Wait()
+	return nil
+}
+
+func (s *SyncService) HandleBlockAnnounceHandshake(from peer.ID, msg *network.BlockAnnounceHandshake) error {
+	logger.Infof("receiving a block announce handshake: %s", from.String())
+	if err := s.workerPool.fromBlockAnnounceHandshake(from); err != nil {
+		return err
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return s.currentStrategy.OnBlockAnnounceHandshake(from, msg)
+}
+
+func (s *SyncService) HandleBlockAnnounce(from peer.ID, msg *network.BlockAnnounceMessage) error {
+	repChange, err := s.currentStrategy.OnBlockAnnounce(from, msg)
+	if repChange != nil {
+		s.network.ReportPeer(repChange.rep, repChange.who)
+	}
+
+	if err != nil {
+		return fmt.Errorf("while handling block announce: %w", err)
+	}
+
+	return nil
+}
+
+func (s *SyncService) OnConnectionClosed(who peer.ID) {
+	logger.Tracef("removing peer worker: %s", who.String())
+	s.workerPool.removeWorker(who)
+}
+
+func (s *SyncService) CreateBlockResponse(who peer.ID, req *network.BlockRequestMessage) (
```

Comment:
Is this, as well as methods below, not used?

Reply:
I need to bring them from the old sync package